core/shared/src/main/scala/scalaz/zio/stream/Stream.scala

```diff
@@ -0,0 +1,733 @@
+        if (i >= c.length) IO.now(Step.cont(s))
+        else
+          f(s, c(i)) flatMap {
+            case Step.Cont(s) => loop(s, c, i + 1)
+            case s            => IO.now(s)
+          }
+
+      self.fold[E1, A, S](s)((s, a) => loop(s, f0(a), 0))
+    }
+  }
+
+  /**
+   * Maps over elements of the stream with the specified effectful function.
+   */
+  final def mapM[E1 >: E, B](f0: A => IO[E1, B]): Stream[E1, B] = new Stream[E1, B] {
+    override def fold[E2 >: E1, B1 >: B, S](s: S)(f: (S, B1) => IO[E2, Step[S]]): IO[E2, Step[S]] =
+      self.fold[E2, A, S](s)((s, a) => f0(a).flatMap(f(s, _)))
+  }
+
+  /**
+   * Merges this stream and the specified stream together.
+   */
+  final def merge[E1 >: E, A1 >: A](that: Stream[E1, A1], capacity: Int = 1): Stream[E1, A1] =
+    self.mergeWith(that, capacity)(identity, identity)
+
+  /**
+   * Merges this stream and the specified stream together to produce a stream of
+   * eithers.
+   */
+  final def mergeEither[E1 >: E, B](that: Stream[E1, B], capacity: Int = 1): Stream[E1, Either[A, B]] =
+    self.mergeWith(that, capacity)(Left(_), Right(_))
+
+  /**
+   * Merges this stream and the specified stream together to a common element
+   * type with the specified mapping functions.
+   */
+  final def mergeWith[E1 >: E, B, C](that: Stream[E1, B], capacity: Int = 1)(l: A => C, r: B => C): Stream[E1, C] =
+    new Stream[E1, C] {
+      override def fold[E2 >: E1, C1 >: C, S](s: S)(f: (S, C1) => IO[E2, Step[S]]): IO[E2, Step[S]] = {
+        type Elem = Either[Take[E2, A], Take[E2, B]]
+
+        def loop(leftDone: Boolean, rightDone: Boolean, s: S, queue: Queue[Elem]): IO[E2, Step[S]] =
+          queue.take.flatMap {
+            case Left(Take.Fail(e))  => IO.fail(e)
+            case Right(Take.Fail(e)) => IO.fail(e)
+            case Left(Take.End) =>
+              if (rightDone) IO.now(Step.Cont(s))
+              else loop(true, rightDone, s, queue)
+            case Left(Take.Value(a)) =>
+              f(s, l(a)).flatMap {
+                case Step.Cont(s)     => loop(leftDone, rightDone, s, queue)
+                case s @ Step.Stop(_) => IO.now(s)
+              }
+            case Right(Take.End) =>
+              if (leftDone) IO.now(Step.Cont(s))
+              else loop(leftDone, true, s, queue)
+            case Right(Take.Value(b)) =>
+              f(s, r(b)).flatMap {
+                case Step.Cont(s)     => loop(leftDone, rightDone, s, queue)
+                case s @ Step.Stop(_) => IO.now(s)
+              }
+          }
+
+        (for {
+          queue  <- Queue.bounded[Elem](capacity)
+          putL   = (a: A) => queue.offer(Left(Take.Value(a))).void
+          putR   = (b: B) => queue.offer(Right(Take.Value(b))).void
+          catchL = (e: E2) => queue.offer(Left(Take.Fail(e)))
+          catchR = (e: E2) => queue.offer(Right(Take.Fail(e)))
+          endL   = queue.offer(Right(Take.End)).forever
```

Comment:
Shouldn't it be `endL   = queue.offer(Left(Take.End)).forever` ?

Reply:
Yes, thank you!